[pytorch][triton] flex attention fwd kernel with TMA loads (#151923)

mandroid6 opened the pull request:
Summary:

Device side TMA for flex_attention fwd kernel, Q K V tensors

Test Plan:
Unit test:
```
buck test 'fbcode//mode/opt' fbcode//caffe2/test/inductor:flex_attention -- test_tma_with_customer_kernel_options
```
https://www.internalfb.com/intern/testinfra/testrun/14355223891618726

Differential Revision: D71082691




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

mandroid6 commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -1546,6 +1651,10 @@ def flex_attention(
             mutated_inputs=[
                 logsumexp,
             ],
+            workspace_arg=get_tma_workspace_arg(
```
Nice catch!

Updated with `num_programs` corresponding to the input query dimensions.

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -1545,6 +1655,16 @@ def flex_attention(
+                num_tma_descriptors=3,
+                device=query.get_device(),
+                num_programs=(
+                    (num_heads + cur_kernel_options["BLOCK_M"] - 1)
```
can we just call into: https://github.com/pytorch/pytorch/blob/4e1d4333f7a6c9335f1f6d27cc1043b0428c0e42/torch/_inductor/kernel/flex_attention.py#L100

so that they there isn't any future delta

mandroid6 replied:
updated

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -374,6 +375,46 @@ def load_checked_2d(
+    desc_v = None
+    if USE_TMA:
+        TMA_SIZE = 128
+        workspace_base = ws_ptr + q_start * 3 * TMA_SIZE
```
we need to ensure the descriptor base is unique. We have 
B * H * Q/BLOCK_M programs
and then 3 per descriptors per program
So maybe we think about this as a 

B*H x  Q/BLOCK_M  x 3*TMA_size Array

So something like 
desc_m_stride = 3 * TMA_SIZE
BH_stride = tl.num_programs(0) * desc_m_stride

BH_NUM = program_id(1)
workspace_base = ws_ptr + BH_NUM * BH_STRIDE + q_start *desc_m_stride 
desc_q ...

S

mandroid6 replied:
In essence each program will have a unique set descriptors so there isn't any race condition to access the same TMA memory across batches.

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -628,7 +716,18 @@ def forward_block_mn(
+                desc_k,
+                [start_n.to(tl.int32) , kv_start],
+                [BLOCK_N, QK_HEAD_DIM_ROUNDED],
+                Q.dtype.element_ty,
```
do we only need to pass q_along for the dtype? can we update the func to pass that along since I was confused if we needed the base Q again which was suprrising

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -1542,12 +1649,30 @@ def flex_attention(
+                num_tma_descriptors=3,
+                device=query.get_device(),
+                num_programs=(
+                    (seq_len + cur_kernel_options["BLOCK_M"] - 1)
```
nit can we re-ruse the grid_func from above to make sure they stay in check

mandroid6 replied:
Updated

Commit:
[pytorch][triton] flex attention fwd kernel with TMA loads (#152460)

Summary:


Device side TMA for flex_attention fwd kernel, Q K V tensors

NOTE: Default state is USE_TMA=False, currently enablement allowed only through explicit `kernel_options`

Test Plan:
Unit test:
```
buck test 'fbcode//mode/opt' fbcode//caffe2/test/inductor:flex_attention -- test_tma_with_customer_kernel_options
```
https://www.internalfb.com/intern/testinfra/testrun/6755399706997966

Reviewed By: htyu, drisspg

Differential Revision: D71082691
